xarray/core/dataset.py

```diff
@@ -3331,20 +3331,22 @@ def broadcast_like(
 
         dims_map, common_coords = _get_broadcast_dims_map_common_coords(args, exclude)
 
+        # Currently requires a `cast` and returns `Dataset` rather than `Self`, haven't
+        # worked through exactly why
         return _broadcast_helper(
-            cast("T_Dataset", args[1]), exclude, dims_map, common_coords
+            cast("Dataset", args[1]), exclude, dims_map, common_coords
```

Comment:
I guess we have to look at what align does to subclasses

Reply:
Yes, I looked at this, but eventually thought it wasn't worth gating on — it's quite complicated. It's something like "the 'biggest' class" — so "da + da = da", but "da + ds = ds"...

Reply:
I adjusted these a bit, see what you think. Allowed removing some `cast`s